File: app/src/main/java/com/termux/styling/TermuxStyleActivity.java
package com.termux.styling;

import android.app.Activity;
import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.text.SpannableString;
import android.text.method.LinkMovementMethod;
import android.text.util.Linkify;
import android.util.Log;
import android.util.Pair;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.WindowManager;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

public class TermuxStyleActivity extends Activity {

    private static final String DEFAULT_FILENAME = "Default";

    static class Selectable {
        final String displayName;
        final String fileName;

        public Selectable(final String fileName) {
            String name = fileName.replace('-', ' ');
            int dotIndex = name.lastIndexOf('.');
            if (dotIndex != -1) name = name.substring(0, dotIndex);

            this.displayName = capitalize(name);
            this.fileName = fileName;
        }

        @Override
        public String toString() {
            return displayName;
        }
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        // Avoid dim behind:
        getWindow().clearFlags(WindowManager.LayoutParams.FLAG_DIM_BEHIND);
        setContentView(R.layout.layout);

        final Button colorSpinner = (Button) findViewById(R.id.color_spinner);
        final Button fontSpinner = (Button) findViewById(R.id.font_spinner);

        final ArrayAdapter<Selectable> colorAdapter = new ArrayAdapter<>(this, android.R.layout.simple_spinner_dropdown_item);

        colorSpinner.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                final AlertDialog dialog = new AlertDialog.Builder(TermuxStyleActivity.this).setAdapter(colorAdapter, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        copyFile(colorAdapter.getItem(which), true);
                    }
                }).create();
                dialog.setOnShowListener(new DialogInterface.OnShowListener() {
                    @Override
                    public void onShow(DialogInterface dialogInterface) {
                        ListView lv = dialog.getListView();
                        lv.setOnItemLongClickListener(new AdapterView.OnItemLongClickListener() {
                            @Override
                            public boolean onItemLongClick(AdapterView<?> adapterView, View view, int position, long id) {
                                showLicense(colorAdapter.getItem(position), true);
                                return true;
                            }
                        });
                    }
                });
                dialog.show();
            }
        });

        final ArrayAdapter<Selectable> fontAdapter = new ArrayAdapter<>(this, android.R.layout.simple_spinner_dropdown_item);
        fontSpinner.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                final AlertDialog dialog = new AlertDialog.Builder(TermuxStyleActivity.this).setAdapter(fontAdapter, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        copyFile(fontAdapter.getItem(which), false);
                    }
                }).create();
                dialog.setOnShowListener(new DialogInterface.OnShowListener() {
                    @Override
                    public void onShow(DialogInterface dialogInterface) {
                        ListView lv = dialog.getListView();
                        lv.setOnItemLongClickListener(new AdapterView.OnItemLongClickListener() {
                            @Override
                            public boolean onItemLongClick(AdapterView<?> adapterView, View view, int position, long id) {
                                showLicense(fontAdapter.getItem(position), false);
                                return true;
                            }
                        });
                    }
                });
                dialog.show();
            }
        });

        List<Selectable> colorList = new ArrayList<>();
        List<Selectable> fontList = new ArrayList<>();
        for (String assetType : new String[]{"colors", "fonts"}) {
            boolean isColors = assetType.equals("colors");
            String assetsFileExtension = isColors ? ".properties" : ".ttf";
            List<Selectable> currentList = isColors ? colorList : fontList;

            currentList.add(new Selectable(isColors ? DEFAULT_FILENAME : DEFAULT_FILENAME));

            try {
                for (String f : getAssets().list(assetType)) {
                    if (f.endsWith(assetsFileExtension)) currentList.add(new Selectable(f));
                }
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }
        Pair<List<Selectable>, List<Selectable>> result = Pair.create(colorList, fontList);

        colorAdapter.addAll(result.first);
        fontAdapter.addAll(result.second);
    }

    void showLicense(Selectable mCurrentSelectable, boolean colors) {
        try {
            final String assetsFolder = colors ? "colors" : "fonts";
            String fileName = mCurrentSelectable.fileName;
            int dotIndex = fileName.lastIndexOf('.');
            if (dotIndex != -1) fileName = fileName.substring(0, dotIndex);
            fileName += ".txt";

            try (InputStream in = getAssets().open(assetsFolder + "/" + fileName)) {
                byte[] buffer = new byte[in.available()];
                in.read(buffer);
                final SpannableString license = new SpannableString(new String(buffer));
                Linkify.addLinks(license, Linkify.ALL);
                AlertDialog dialog = new AlertDialog.Builder(this)
                        .setTitle(mCurrentSelectable.displayName)
                        .setMessage(license)
                        .setPositiveButton(android.R.string.ok, null)
                        .show();
                ((TextView) dialog.findViewById(android.R.id.message)).setMovementMethod(LinkMovementMethod.getInstance());
            }
        } catch (IOException e) {
            // Ignore.
        }
    }

    void copyFile(Selectable mCurrentSelectable, boolean colors) {
        final String outputFile = colors ? "colors.properties" : "font.ttf";
        try {
            final String assetsFolder = colors ? "colors" : "fonts";

            Context context = createPackageContext("com.termux", Context.CONTEXT_IGNORE_SECURITY);
            File homeDir = new File(context.getFilesDir(), "home");
            File termuxDir = new File(homeDir, ".termux");
            File destinationFile = new File(termuxDir, outputFile);
            if (!(termuxDir.isDirectory() || termuxDir.mkdirs()))
                throw new RuntimeException("Cannot create termux dir=" + termuxDir.getAbsolutePath());

            // Fix for if the user has messed up with chmod:
            destinationFile.setWritable(true);
            destinationFile.getParentFile().setWritable(true);
            destinationFile.getParentFile().setExecutable(true);

            boolean defaultChoice = mCurrentSelectable.fileName.equals(DEFAULT_FILENAME);
            // Write to existing file to keep symlink if this is used.
            try (FileOutputStream out = new FileOutputStream(destinationFile)) {
                if (defaultChoice) {
                    if (colors) {
                        byte[] comment = "# Using default color theme.".getBytes(StandardCharsets.UTF_8);
                        out.write(comment);
                    } else {
                        // Just leave an empty font file as a marker.
                    }
                } else {
                    try (InputStream in = getAssets().open(assetsFolder + "/" + mCurrentSelectable.fileName)) {
                        byte[] buffer = new byte[4096];
                        int len;
                        while ((len = in.read(buffer)) > 0)
                            out.write(buffer, 0, len);
                    }
                }
            }

            // Note: Must match constant in Term#onCreate():
            final String ACTION_RELOAD = "com.termux.app.reload_style";
            Intent executeIntent = new Intent(ACTION_RELOAD);
            executeIntent.putExtra(ACTION_RELOAD, colors ? "colors" : "font");
            sendBroadcast(executeIntent);
        } catch (Exception e) {
            Log.w("termux", "Failed to write " + outputFile, e);
            String message = getResources().getString(R.string.writing_failed) + e.getMessage();
            Toast.makeText(this, message, Toast.LENGTH_LONG).show();
        }
    }

    private static String capitalize(String str) {
        boolean lastWhitespace = true;
        char[] chars = str.toCharArray();
        for (int i = 0; i < chars.length; i++) {
            if (Character.isLetter(chars[i])) {
                if (lastWhitespace) chars[i] = Character.toUpperCase(chars[i]);
                lastWhitespace = false;
            } else {
                lastWhitespace = Character.isWhitespace(chars[i]);
            }
        }
        return new String(chars);
    }
}

File: app/src/main/java/com/termux/styling/TermuxStyleActivity.kt
package com.termux.styling

import android.app.Activity
import android.app.AlertDialog
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.text.SpannableString
import android.text.method.LinkMovementMethod
import android.text.util.Linkify
import android.util.AtomicFile
import android.util.Log
import android.view.View
import android.view.WindowManager
import android.widget.ArrayAdapter
import android.widget.Button
import android.widget.TextView
import android.widget.Toast
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.nio.charset.StandardCharsets
import java.util.*

const val DEFAULT_FILENAME = "Default"

fun capitalize(str: String): String {
    var lastWhitespace = true
    val chars = str.toCharArray()
    for (i in chars.indices) {
        if (Character.isLetter(chars[i])) {
            if (lastWhitespace) chars[i] = Character.toUpperCase(chars[i])
            lastWhitespace = false
        } else {
            lastWhitespace = Character.isWhitespace(chars[i])
        }
    }
    return String(chars)
}

class TermuxStyleActivity : Activity() {

    internal class Selectable(val fileName: String) {
        val displayName: String

        init {
            var name = fileName.replace('-', ' ')
            val dotIndex = name.lastIndexOf('.')
            if (dotIndex != -1) name = name.substring(0, dotIndex)

            this.displayName = capitalize(name)
        }

        override fun toString(): String {
            return displayName
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        // Avoid dim behind:
        window.clearFlags(WindowManager.LayoutParams.FLAG_DIM_BEHIND)
        setContentView(R.layout.layout)

        val colorSpinner = findViewById<Button>(R.id.color_spinner)
        val fontSpinner = findViewById<Button>(R.id.font_spinner)

        val colorAdapter = ArrayAdapter<Selectable>(this, android.R.layout.simple_spinner_dropdown_item)

        colorSpinner.setOnClickListener {
            val dialog = AlertDialog.Builder(this@TermuxStyleActivity).setAdapter(colorAdapter) { _, which -> copyFile(colorAdapter.getItem(which), true) }.create()
            dialog.setOnShowListener {
                val lv = dialog.listView
                lv.setOnItemLongClickListener { _, _, position, _ ->
                    showLicense(colorAdapter.getItem(position) as Selectable, true)
                    true
                }
            }
            dialog.show()
        }

        val fontAdapter = ArrayAdapter<Selectable>(this, android.R.layout.simple_spinner_dropdown_item)
        fontSpinner.setOnClickListener {
            val dialog = AlertDialog.Builder(this@TermuxStyleActivity).setAdapter(fontAdapter) { _, which -> copyFile(fontAdapter.getItem(which), false) }.create()
            dialog.setOnShowListener {
                val lv = dialog.listView
                lv.setOnItemLongClickListener { _, _, position, _ ->
                    showLicense(fontAdapter.getItem(position) as Selectable, false)
                    true
                }
            }
            dialog.show()
        }

        val colorList = ArrayList<Selectable>()
        val fontList = ArrayList<Selectable>()
        for (assetType in arrayOf("colors", "fonts")) {
            val isColors = assetType == "colors"
            val assetsFileExtension = if (isColors) ".properties" else ".ttf"
            val currentList = if (isColors) colorList else fontList

            currentList.add(Selectable(if (isColors) DEFAULT_FILENAME else DEFAULT_FILENAME))

            try {
                assets.list(assetType)!!
                        .filter { it.endsWith(assetsFileExtension) }
                        .forEach { currentList.add(Selectable(it)) }
            } catch (e: IOException) {
                throw RuntimeException(e)
            }

        }

        colorAdapter.addAll(colorList)
        fontAdapter.addAll(fontList)
    }

    private fun showLicense(mCurrentSelectable: Selectable, colors: Boolean) {
        try {
            val assetsFolder = if (colors) "colors" else "fonts"
            var fileName = mCurrentSelectable.fileName
            val dotIndex = fileName.lastIndexOf('.')
            if (dotIndex != -1) fileName = fileName.substring(0, dotIndex)
            fileName += ".txt"

            assets.open("$assetsFolder/$fileName").use { `in` ->
                val buffer = ByteArray(`in`.available())
                `in`.read(buffer)
                val license = SpannableString(String(buffer))
                Linkify.addLinks(license, Linkify.ALL)
                val dialog = AlertDialog.Builder(this)
                        .setTitle(mCurrentSelectable.displayName)
                        .setMessage(license)
                        .setPositiveButton(android.R.string.ok, null)
                        .show()
                (dialog.findViewById<View>(android.R.id.message) as TextView).movementMethod = LinkMovementMethod.getInstance()
            }
        } catch (e: IOException) {
            // Ignore.
        }

    }

    private fun copyFile(mCurrentSelectable: Selectable?, colors: Boolean) {
        val outputFile = if (colors) "colors.properties" else "font.ttf"
        try {
            val assetsFolder = if (colors) "colors" else "fonts"

            val context = createPackageContext("com.termux", Context.CONTEXT_IGNORE_SECURITY)
            val homeDir = File(context.filesDir, "home")
            val termuxDir = File(homeDir, ".termux")
            if (!(termuxDir.isDirectory || termuxDir.mkdirs()))
                throw RuntimeException("Cannot create termux dir=" + termuxDir.absolutePath)

            // Use canonicalFile to follow a possible symlink:
            val destinationFile = File(termuxDir, outputFile).canonicalFile
            // Fix for if the user has messed up with chmod:
            destinationFile.setWritable(true)
            destinationFile.parentFile.setWritable(true)
            destinationFile.parentFile.setExecutable(true)

            val defaultChoice = mCurrentSelectable!!.fileName == DEFAULT_FILENAME
            val atomicFile = AtomicFile(destinationFile)
            val out = atomicFile.startWrite()
            if (defaultChoice) {
                if (colors) {
                    val comment = "# Using default color theme.".toByteArray(StandardCharsets.UTF_8)
                    out.write(comment)
                } else {
                    // Just leave an empty font file as a marker.
                }
            } else {
                assets.open(assetsFolder + "/" + mCurrentSelectable.fileName).use {
                    it.copyTo(out)
                }
            }
            atomicFile.finishWrite(out)

            // Note: Must match constant in Term#onCreate():
            val ACTION_RELOAD = "com.termux.app.reload_style"
            val executeIntent = Intent(ACTION_RELOAD)
            executeIntent.putExtra(ACTION_RELOAD, if (colors) "colors" else "font")
            sendBroadcast(executeIntent)
        } catch (e: Exception) {
            Log.w("termux", "Failed to write $outputFile", e)
            val message = resources.getString(R.string.writing_failed) + e.message
            Toast.makeText(this, message, Toast.LENGTH_LONG).show()
        }

    }

}
